feat: allow force-destroying a node in booting state

utkuozdemir opened the pull request:
Implement force-destroy logic to allow machines to be reset and removed from the cluster even if they are stuck in the booting phase. This is useful in cases like a node not being able to join to etcd, and it is known by the cluster operator that it will never be able to.

Closes https://github.com/siderolabs/omni/issues/779.

Unix4ever commented on internal/backend/runtime/omni/state_validation.go:
```diff
@@ -1120,3 +1120,20 @@ func infraMachineConfigValidationOptions(st state.State) []validated.StateOption
 	}
 }
+
+func nodeForceDestroyRequestValidationOptions(st state.State) []validated.StateOption {
```
We probably need the unit test for that.

utkuozdemir replied:
good point, added

Unix4ever commented on internal/backend/runtime/omni/controllers/omni/cluster_machine_config_status_test.go:
```diff
@@ -384,6 +384,93 @@ func (suite *ClusterMachineConfigStatusSuite) TestResetUngraceful() {
+		suite.Require().NoError(err)
+	}
+
+	time.Sleep(time.Second * 1)
```
Can we avoid having this timeout here? Maybe some more reliable way to wait?

utkuozdemir replied:
actually it was not needed, so I removed it

Unix4ever commented on internal/backend/runtime/omni/controllers/omni/cluster_machine_config_status.go:
```diff
@@ -689,6 +704,18 @@ func (h *clusterMachineConfigStatusControllerHandler) reset(
 		graceful = false
 	}
 
+	nodeForceDestroyRequest, err := safe.ReaderGetByID[*omni.NodeForceDestroyRequest](ctx, h.r, machineConfig.Metadata().ID())
+	if err != nil && !state.IsNotFoundError(err) {
+		return fmt.Errorf("failed to get node force delete request %q: %w", machineConfig.Metadata().ID(), err)
+	}
+
+	forceDestroy := nodeForceDestroyRequest != nil
+	if forceDestroy {
+		logger.Info("node is requested to be force destroyed")
+
+		graceful = false
+	}
```
Maybe lets extract that to a separate method?

utkuozdemir replied:
good idea, done

Unix4ever reviewed (commented):
I'd also suggest to add CLI commands to force destroy the node.

utkuozdemir commented:
> I'd also suggest to add CLI commands to force destroy the node.

Done, just added `omnictl cluster machine delete <id> [--force]`

While doing it, I moved the logic of the `omnictl delete <restype> <id>` command to a common place, its own package, and started using it from both commands.

One additional thing: I made this destroy command to log human-readable logs to `stderr` instead of previously `stdout`. I think `stderr` is actually the right target for it, but let me know if you think otherwise - currently we have both in different commands and are not consistent. (I even have better ideas for our CLIs outputs).

utkuozdemir commented:
/m
